Date: Wed, 28 May 2014 18:00:59 -0700
From: Philip Guenther <pguenther@proofpoint.com>
To: hackers@openbsd.org
Subject: perl: stop using /dev/arandom
User-Agent: Alpine 2.11 (BSO 23 2013-08-11)


{Perl_,}seed() is a maybe-public API in libperl (c.f. perlapi(1)) which
currently reads a U32 from /dev/arandom.  That's not as good as just
calling arc4random(), as the device doesn't work in chroots, etc.  While
it's configurable for various other sources, Perl_seed() doesn't currently
have a "just call this other function, dammit" define, so here's the blunt
hammer diff to pound arc4random() into it.

(Perl_seed() is only used in perl itself as the seed generator if you call
perl-level srand() without an argument, and similarly the perl-level
rand() without previously seeding it.)

Yay/nay?

--- ./util.c.orig	Mon Jun 22 07:48:01 2026
+++ ./util.c	Tue Aug 25 07:33:33 2026
@@ -4669,6 +4669,9 @@ Perl_seed(pTHX)
 {
     PERL_ARGS_ASSERT_SEED;
 
+#if defined(__OpenBSD__)
+    return arc4random();
+#else
    /*
     * Attempt to read from /dev/urandom to generate a pseudo-random number.
     * If that does not work, or it is unavailable, we fall back to gathering
@@ -4743,6 +4746,7 @@ Perl_seed(pTHX)
     /* PerlIO_printf(Perl_debug_log, "SEED: %lu\n", ret); */
 
     return ret;
+#endif
 }
 
 void
